test/Test/TH.hs

```diff
@@ -39,13 +41,41 @@ instance TestEquality T2Repr where
                     [ (AnyType, [|testEquality|]) ])
 deriving instance Show (T2Repr t)
 
+data T3 (is_a :: Symbol) where
+  T3_Int :: Int -> T3 "int"
+  T3_Bool :: Bool -> T3 "bool"
+$(return [])
+instance TestEquality T3 where
+  testEquality = $(structuralTypeEquality [t|T3|] [])
+instance Eq (T3 s) where
+  (==) = $(structuralEquality [t|T3|] [])
+deriving instance Show (T3 s)
+
+data T4 b (is_a :: Symbol) where
+  T4_Int :: Int -> T4 b "int"
+  T4_Bool :: Bool -> T4 b "bool"
+$(return [])
+instance TestEquality (T4 b) where
+  testEquality = $(structuralTypeEquality [t|T4|] [])
+instance Eq (T4 b s) where
+  (==) = $(structuralEquality [t|T4|] [])
+deriving instance Show (T4 b s)
+
 eqTest :: (TestEquality f, Show (f a), Show (f b)) => f a -> f b -> IO ()
 eqTest a b =
   when (not (isJust (testEquality a b))) $ assertFailure $ show a ++ " /= " ++ show b
 
 neqTest :: (TestEquality f, Show (f a), Show (f b)) => f a -> f b -> IO ()
 neqTest a b =
-  when (isJust (testEquality a b)) $ assertFailure $ show a ++ " == " ++ show b
+  when (isJust (testEquality a b))
+  $ assertFailure
+  $ show a <> " == " <> show b <> " but should not be!"
```

Comment:
nit: Looks like these are still strings, so I'd prefer keeping `++` rather than `<>`, as it's less polymorphic, and so lets the reader know the more specific types involved.